File: android/mobilepasssdk/src/main/java/com/armongate/mobilepasssdk/service/AnalyticsService.java
package com.armongate.mobilepasssdk.service;

import android.content.Context;

import com.armongate.mobilepasssdk.constant.LogCodes;
import com.armongate.mobilepasssdk.constant.StorageKeys;
import com.armongate.mobilepasssdk.manager.LogManager;
import com.armongate.mobilepasssdk.manager.StorageManager;
import com.armongate.mobilepasssdk.model.AnalyticsQueueItem;
import com.armongate.mobilepasssdk.model.request.RequestAnalyticsData;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;

public class AnalyticsService {
    private static final int MAX_RETRY_AGE_HOURS = 72;

    private List<AnalyticsQueueItem> loadQueue(Context context) {
        try {
            String data = StorageManager.getInstance().getValue(context, StorageKeys.ANALYTICS_QUEUE);
            Type listType = new TypeToken<ArrayList<AnalyticsQueueItem>>(){}.getType();
            List<AnalyticsQueueItem> queue = new Gson().fromJson(data, listType);

            if (queue == null) return new ArrayList<>();
            
            // Filter out old entries
            long cutoffTime = System.currentTimeMillis() - (MAX_RETRY_AGE_HOURS * 3600 * 1000);
            List<AnalyticsQueueItem> filteredQueue = new ArrayList<>();
            for (AnalyticsQueueItem item : queue) {
                if (item.timestamp > cutoffTime) {
                    filteredQueue.add(item);
                }
            }
            return filteredQueue;

        } catch (Exception e) {
            LogManager.getInstance().error("Failed to load analytics queue: " + e.getLocalizedMessage(), LogCodes.OTHER);
            return new ArrayList<>();
        }
    }
    
    private void saveQueue(Context context, List<AnalyticsQueueItem> queue) {
        try {
            String data = new Gson().toJson(queue);
            StorageManager.getInstance().setValue(context, StorageKeys.ANALYTICS_QUEUE, data);
        } catch (Exception e) {
            LogManager.getInstance().error("Failed to save analytics queue: " + e.getLocalizedMessage(), LogCodes.OTHER);
        }
    }

    public void sendAnalytics(Context context, RequestAnalyticsData request, BaseService.ServiceResultListener listener) {
        // Load and clear existing queue
        List<AnalyticsQueueItem> queue = loadQueue(context);
        List<AnalyticsQueueItem> queuedItems = new ArrayList<>(queue);
        queue.clear();
        saveQueue(context, queue);

        // Add new request to queued items
        queuedItems.add(new AnalyticsQueueItem(request));

        List<AnalyticsQueueItem> failedItems = new ArrayList<>();
        AtomicInteger completeCount = new AtomicInteger(0);

        // Process all queued items
        for (AnalyticsQueueItem item : queuedItems) {
            BaseService.getInstance().requestPost("api/v2/analytics", item.request, null, new BaseService.ServiceResultListener() {
                @Override
                public void onCompleted(Object result) {
                    processCompletion(context, queuedItems, failedItems, completeCount, item, null);
                }
                
                @Override
                public void onError(int statusCode, String message) {
                    processCompletion(context, queuedItems, failedItems, completeCount, item, message);
                }
            });
        }
    }

    private synchronized void processCompletion(Context context, 
                                              List<AnalyticsQueueItem> queuedItems,
                                              List<AnalyticsQueueItem> failedItems, 
                                              AtomicInteger completeCount,
                                              AnalyticsQueueItem item,
                                              String errorMessage) {
        // Add to failed items if there was an error and item isn't too old
        if (errorMessage != null) {
            long cutoffTime = System.currentTimeMillis() - (MAX_RETRY_AGE_HOURS * 3600 * 1000);
            if (item.timestamp > cutoffTime) {
                failedItems.add(item);
            }
        }

        // Check if all items are processed
        if (completeCount.incrementAndGet() == queuedItems.size()) {
            if (!failedItems.isEmpty()) {
                // Load current queue
                List<AnalyticsQueueItem> currentQueue = loadQueue(context);
                
                // Add failed items
                currentQueue.addAll(failedItems);
                
                // Save updated queue
                saveQueue(context, currentQueue);
            }
        }
    }
}

File: ios/MobilePassSDK/Service/AnalyticsService.swift
//
//  AnalyticsService.swift
//  MobilePassSDK
//
//  Created by Erinc Cakir on 4.03.2025.
//


class AnalyticsService {
    private enum Constants {
        static let maxRetryAgeHours: TimeInterval = 72
        static let analyticsQueueKey = StorageKeys.ANALYTICS_QUEUE
    }

    private func loadQueue() -> [AnalyticsQueueItem] {
        guard let data = try? StorageManager.shared.getValue(key: Constants.analyticsQueueKey, secure: false),
                let jsonData = data.data(using: .utf8),
                let queue = try? JSONDecoder().decode([AnalyticsQueueItem].self, from: jsonData) else {
            return []
        }
        
        let cutoffDate = Date().addingTimeInterval(-Constants.maxRetryAgeHours * 3600)
        return queue.filter { $0.timestamp > cutoffDate }
    }
    
    private func saveQueue(_ queue: [AnalyticsQueueItem]) {
        do {
            let data = try JSONEncoder().encode(queue)
            guard let str = String(data: data, encoding: .utf8) else {
                LogManager.shared.error(message: "Failed to encode analytics queue")
                return
            }
            try StorageManager.shared.setValue(key: Constants.analyticsQueueKey, value: str, secure: false)
        } catch {
            LogManager.shared.error(message: "Failed to save analytics queue: \(error)")
        }
    }

    func sendAnalytics(request: RequestAnalyticsData, completion: @escaping (Result<ResponseDefault?, RequestError>) -> Void) {
        var queue = loadQueue()
        var queuedItems = queue
        queue.removeAll()
        saveQueue(queue)

        queuedItems.append(AnalyticsQueueItem(request: request))

        var failedItems: [AnalyticsQueueItem] = []
        var completeCount = 0;

        // Process queued items first
        for item in queuedItems {
            BaseService.shared.requestPost(url: "api/v2/analytics", data: item.request.dictionary) { [weak self] (result: Result<ResponseDefault?, RequestError>) in
                if case .failure(_) = result {
                    // Re-queue failed item if it's not too old
                    if item.timestamp > Date().addingTimeInterval(-Double(Constants.maxRetryAgeHours * 3600)) {
                        failedItems.append(item)
                    }
                }

                completeCount += 1
                
                if (completeCount == queuedItems.count) {
                    if !failedItems.isEmpty {
                        var currentQueue: [AnalyticsQueueItem] = []

                        // Load queue
                        if let data = try? StorageManager.shared.getValue(key: Constants.analyticsQueueKey, secure: false),
                           let queue = try? JSONDecoder().decode([AnalyticsQueueItem].self, from: data.data(using: .utf8)!) {
                            let cutoffDate = Date().addingTimeInterval(-Double(Constants.maxRetryAgeHours * 3600))
                            currentQueue = queue.filter { $0.timestamp > cutoffDate }
                        }

                        // Append failed items to queue
                        currentQueue.append(contentsOf: failedItems)

                        // Save queue
                        if let data = try? JSONEncoder().encode(currentQueue),
                           let str = String(data: data, encoding: .utf8) {
                            _ = try? StorageManager.shared.setValue(key: Constants.analyticsQueueKey, value: str, secure: false)
                        }
                    }
                }
            }
        }
    }
    
}
